### PR TITLE
Make HAB_STUDIO_BINARY env var override work to run alternate studio …

### PR DESCRIPTION
**Pull request by baumanj**
…code

This will now successfully set up a new studio as well as allow a previously
created studio to be entered while running new studio code. The prompt will
be updated to reflect the studio code being run.

See https://github.com/habitat-sh/habitat/issues/4531

Signed-off-by: Jon Bauman <5906042+baumanj@users.noreply.github.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by fnichol** [comment]
I'd like to test this out first to understand what I would see on the prompt. On first look, this would be useful although I'm not sold on outputing more information to the prompt (we don't do this for overriding Supervisor behavior, using origin keys, etc.). I should have a chance mid-day to build and test this and will update.

**Comment by baumanj** [comment]
Here's what it would look like, @fnichol:
```
➤ hab studio enter
…
[1][default:/src:0]# 
```
```
➤ env HAB_STUDIO_BINARY=$HOME/habitat/components/studio/bin/hab-studio.sh hab studio enter
…
[1][HAB_STUDIO_BINARY:/home/jbauman/habitat/components/studio/bin/hab-studio.sh][default:/src:0]# 
```
I'm definitely open to rethinking or removing the changes to the prompt, but what about the core functionality for making the `HAB_STUDIO_BINARY` override work? That is: using `hab studio version` and `hab pkg path` to resolve the necessary `version` and `libexec_path` variables. Is relying on the installed `hab` for those things appropriate?

File: components/studio/bin/hab-studio.sh [hunk 1]
```diff
@@ -991,6 +991,11 @@ chroot_env() {
   if [ -n "${HAB_UPDATE_STRATEGY_FREQUENCY_MS:-}" ]; then
     env="$env HAB_UPDATE_STRATEGY_FREQUENCY_MS=$HAB_UPDATE_STRATEGY_FREQUENCY_MS"
   fi
+  # If a Habitat studio binary is set, then propagate it into the Studio's
+  # environment.
+  if [ -n "${HAB_STUDIO_BINARY:-}" ]; then
```
Comment (christophermaier) [inline components/studio/bin/hab-studio.sh]: How will this behave if `HAB_STUDIO_BINARY` is given as an absolute path (i.e., a binary that exists at a path on the host system that's not at that location inside the studio)? 

Is that a valid concern? I think I've lost track at which Inception level we may be on at this point in the code :smile:

**Comment by christophermaier** [comment]
@baumanj I think the overall approach is probably sound, though I haven't yet run it to try and break it 😄 

As for the prompt, maybe we could just add a single glyph of some kind to indicate that you're running with an overridden studio, and then you can consult your `env` to discover the concrete details. That would provide a useful visual indication of being in a non-standard state, while not eating up a large portion of the line with prompt.

File: components/studio/bin/hab-studio.sh (hunk 1 above)
Comment (baumanj) [reply components/studio/bin/hab-studio.sh]: I've actually only tested it with `HAB_STUDIO_BINARY` set to an absolute path:
```
11:13:10 AM jbauman@ubuntu:~/some-non-hab-source-dir
➤ env HAB_STUDIO_BINARY=$HOME/habitat/components/studio/bin/hab-studio.sh hab studio enter
```
But it does seem to work if the binary is not at a location inside the studio as above.

Comment (christophermaier) [reply components/studio/bin/hab-studio.sh]: Oh, that makes total sense then; that binary determines what studio is run, so it's by definition relative to the workstation, and not the inside of the studio. 

I was one level of Inception too deep 😆

**Comment by baumanj** [comment]
I considered what single glyph made sense to alert users they were using an overridden studio binary and ultimately decided that a single glyph was too abstruse. But I agree that the whole path is to verbose, so hopefully this is the happy, usable middle path:
```
➤ hab studio enter
…
[1][default:/src:0]# 
…
➤ env HAB_STUDIO_BINARY=$HOME/habitat/components/studio/bin/hab-studio.sh hab studio enter
…
[1][HAB_STUDIO_BINARY][default:/src:0]# echo $HAB_STUDIO_BINARY
/home/jbauman/habitat/components/studio/bin/hab-studio.sh
```

**Comment by christophermaier** [comment]
Hrmm... Just looking at the output, I don't think I'd have guessed that `HAB_STUDIO_BINARY` in the prompt meant that something had been overridden, though to be fair, you'd need extra information looking at a single glyph, too. I'm not sure I have a better suggestion, though! Also, if you're even seeing this, you're already likely to know what you're doing, almost by definition.

I'm 👍 on this. It's great to have an indicator that you're running in a non-standard state, which is incredibly useful on it's own.

**Pull request merged**